### PR TITLE
`Programming exercises`: Add information box to exercise details page

### PR DESCRIPTION
**Pull request by rabeatwork**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR is part of the new design; it enhances usability and aligns with the updated layout.

### Description

All important information for an exercise is now displayed in the information boxes at the top. Additionally, the Instructor Actions are consolidated into a dropdown button.
This is a reopened PR of the stale PR #9089: Programming exercises: Add information box to exercise details page.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to Exercises
3. Check differently configured exercises and verify that the information are displayed correctly 

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| exam-start-information.component.ts | 100% | ✅ |
| exercise-headers-information.component.ts | 76.47% | ✅ |
| result.component.ts | 88.88% | ✅ |
| updating-result.component.ts | 97.05% | ✅ |
| course-overview.component.ts | 87.15% | ✅ |
| submission-result-status.component.ts | 100% | ✅ |
| code-button.component.ts | 93.28% | ✅  |
| difficulty-level.component.ts | 94.11% | ✅  |
| exercise-categories.component.ts | 100% | ✅ |
| information-box.component.ts | 100% | ✅ |
| date.utils.ts | 100% | ✅ |

### Screenshots

Text Future und Kategorie
ModelingCurrent
Programming + Result
Optional Text
QuizBewertet

## Release Notes

- **New Features**
	- Introduced a new `ExerciseHeadersInformationComponent` to display detailed exercise information.
	- Added a `DifficultyLevelComponent` for visual representation of difficulty levels.
	- Enhanced rendering logic for exercise categories and submission results.
	- Implemented a new `InformationBox` structure for better content handling.
	- Added a new input property `showCompletion` to control the display of completion information in results.

- **Improvements**
	- Streamlined the display logic for exercise details and submission policies.
	- Improved localization for submission penalties and titles in multiple languages.
	- Adjusted sidebar and drawer dimensions for better layout consistency.
	- Enhanced the styling of message input areas and card components for improved responsiveness.
	- Updated the navbar and breadcrumb styles for better visual hierarchy and consistency.

- **Bug Fixes**
	- Adjusted conditional rendering for completion dates and exercise action buttons.
	- Fixed rendering issues related to the sidebar and exercise detail components.
	- Resolved issues with the visibility of score displays in exercise results.

- **Styling Updates**
	- Refined styles for various components, including sidebar and card elements, to improve user experience.
	- Updated styles for message input areas and card components for better responsiveness.

- **Tests**
	- Added comprehensive test suites for new components and utility functions to ensure functionality and reliability.
	- Enhanced test coverage for existing components to validate rendering and behavior under various conditions.

### CODE REVIEW

**Review by sachmii — APPROVED** [review]
Tested on TS2, changes look good. Everything works fine.

Bildschirmfoto 2024-11-04 um 12 19 18

File: src/main/webapp/i18n/de/programmingExercise.json [hunk 1]
```diff
@@ -601,10 +601,10 @@
                     },
                     "submission_penalty": {
                         "title": "Punktabzug je Überschreitung",
-                        "tooltip": "Das System zieht für jede Abgabe, die das Abgabelimit überschreitet, Punkte vom Gesamtergebnis ab. Versuche, innerhalb der erlaubten Abgaben zu bleiben, um Abzüge zu vermeiden!"
+                        "tooltip": "Das System zieht für jede Abgabe, die das Abgabelimit überschreitet, Punkte vom Gesamtergebnis ab. Der Punktabzug für diese Aufgabe ist -{{points}} Punkte"
```
Comment (b-fein) [inline src/main/webapp/i18n/de/programmingExercise.json:604]:
```suggestion
                        "tooltip": "Das System zieht für jede Abgabe, die das Abgabelimit überschreitet, Punkte vom Gesamtergebnis ab. Der Punktabzug für diese Aufgabe ist -{{points}} Punkte."
```

File: src/main/webapp/i18n/en/programmingExercise.json [hunk 2]
```diff
@@ -601,10 +601,10 @@
                     },
                     "submission_penalty": {
                         "title": "Submission Penalty",
-                        "tooltip": "The system deducts points from your score for each submission that exceeds the submission limit. Try to stay within the submission limit to avoid penalties!"
+                        "tooltip": "The system deducts points from your score for each submission that exceeds the submission limit. The penalty for this exercise is -{{points}} points"
```
Comment (b-fein) [inline src/main/webapp/i18n/en/programmingExercise.json:604]:
```suggestion
                        "tooltip": "The system deducts points from your score for each submission that exceeds the submission limit. The penalty for this exercise is -{{points}} points."
```

File: src/main/webapp/i18n/en/student-dashboard.json [hunk 3]
```diff
@@ -269,53 +269,54 @@
...
+                "exampleSolutionPublicationDate": "Example solution publication date",
                 "exampleSolutionPublicationDateTooltip": "The example solution is published at {{date}}.",
-                "assessmentDue": "Assessment due:",
+                "assessmentDue": "Assessment due",
                 "assessmentDueTooltip": "The assessment is due by {{date}}.",
-                "complaintDue": "Complaint due:",
-                "complaintDueTooltip": "You can write a complaint until {{date}}.",
-                "complaintPossible": "Complaint possible:",
+                "complaintDue": "Complaint due",
+                "complaintDueTooltip": "You can write a complain until {{date}}.",
```
Comment (b-fein) [inline src/main/webapp/i18n/en/student-dashboard.json:314]:
```suggestion
                "complaintDueTooltip": "You can write a complaint until {{date}}.",
```

**Review by kevinfischer4 — DISMISSED** [review]
Tested on TS2. Looked at different exercise types. All information that is needed is displayed correctly (difficulty, current score, deadline, points/bonus,...). Cool improvement of the GUI :)

**Review by edkaya — COMMENTED** [review]
Tested on ts3, works great :) Maybe one small issue regarding information boxes: I think there is no vertical spacing between boxes anymore, which seems a little bit off if the boxes are displayed one under the other. Maybe it should not be problematic for exercises, but it will be the case in exam mode, where we have at least 7 information boxes. 
Exercises: 
Screenshot 2024-11-04 at 19 19 00
in Exam Mode: 

Screenshot 2024-11-04 at 19 19 19

**Review by vinceclifford — APPROVED** [review]
Tested on TS3. Everything works as expected

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Every information of different exercise types are displayed correctly. Really love the new GUI

**Comment by rabeatwork** [comment]
> Tested on ts3, works great :) Maybe one small issue regarding information boxes: I think there is no vertical spacing between boxes anymore, which seems a little bit off if the boxes are displayed one under the other. Maybe it should not be problematic for exercises, but it will be the case in exam mode, where we have at least 7 information boxes. Exercises: Screenshot 2024-11-04 at 19 19 00 in Exam Mode:
> 
> Screenshot 2024-11-04 at 19 19 19

Good Catch! I'll fix it.

**Review by edkaya — APPROVED** [review]
Tested on ts5, thanks for the fix 👍 Everything seems working fine now

**Review by flbrgit — APPROVED** [review]
Tested on TS2 and everything works as expected, also when boxes are displayed under each other. 
Screenshot 2024-11-05 162316

The tooltips of categories also work as expected. Is there any way to also specify a tooltip of a custom category?

**Review by sachmii — APPROVED** [review]
Tested on TS5, reapprove

**Review by az108 — APPROVED** [review]
Tested on TS5, works without any issues 👍

**Review by HanyangXu0508 — APPROVED** [review]
tested locally and worked as expected.

**Review by ahbitaqu — APPROVED** [review]
[Tested on TS2]
Works as described and expected, couldn't find any issues.
Apart from functionality tough, I think that the labels for the category, bonus, etc. are small and crowded, which might make them difficult to read.
image

**Review by JerroyTan — APPROVED** [review]
[Tested on TS5]
I tested the boxes for Points, Submission Due, Start date, Status, Difficulty and Categories. Functionalities are all working as expected. Boxes also wrap on mobile view
Comment: 
- Difficulty box seems to be less obvious than the previous fully-colored background boxes 
- Box for Categories may be too small: it is very hard to read

**Review by sawys777 — APPROVED** [review]
Tested on TS3, everything worked accordingly to the test description, however the tags in the categories box is a little too small.

**Pull request merged**